### ISSUE TITLE
Make ray() aware of walls

### ISSUE THREAD

**Issue by leftiness** [leftiness/hex_math#41]
I'll need to use a HashMap<Point, HasWalls> instead of a HashSet<Point>. Rename opaque to walls. Use #39 to determine which wall would be passed through when I'm on a line from A to B. Then break if I find a point along the line where a strong enough wall exists.

**Comment by leftiness** [comment]
Given
- a regular hexagon with vertices ABCDEF
- a point P within the hexagon
- a point Q which is outside of the hexagon

The process of determining the hexagon edge through which the line PQ will exit is:
- Determine the three edges AB, BC, CD through which the line PQ could possibly exit. For example, a line PQ which is described as "northeast" can only exit through the northwest, northeast, or east edges no matter where exactly P and Q are.
- Determine the slopes of lines PB and PC.
- If the slope of PQ is greater than slope PB, then the line will exit from edge AB.
- If the slope of PQ is less than line PB and greater than line PC, then the line will exit from edge BC.
- If the slope is less than line PC, then the line will exit from edge CD.

**Comment by leftiness** [comment]
ray_with_strength() will have to return not just a `Set<Point>` but a `Map<Point, i32>` where `strength_of_ray_at_this_point: i32 = result.get(this_point);`

**Comment by leftiness** [comment]
Determining the three edges which could be possible exits is maybe just checking which slope is biggest? I've got (dq, dr, ds, dt). The comparative size of those dQRS determines the direction.

I'll also have to add another step to that process. I have to figure out if the line will exit through the top/bottom instead of the side. I'll have to think about how that will work...

**Comment by leftiness** [comment]
Alternatively.

I know the float point A. That's what I got by adding the step. I know the rounded point B. I get that by rounding my float point. Both of those operations are mandatory.

I could add the value to the rounded point to find the coordinates of the northern vertex of this hex. That's point C. Now I have the triangle ABC. I can find the angle CAB. That one will tell me which wall I'll exit. For example, 0 degrees from origin is the northern vertex. 90 degrees from origin exits from the eastern side.

This approach may be more efficient since I wouldn't have to figure out other slopes and make comparisons. Not sure.

At any rate, I'm still not sure about how I'll determine if I'm exiting through the top/bottom or the side...

**Comment by leftiness** [comment]
Thoughts about determining if exiting top or side.

A hexagon is surrounded by a circle. The hex has an edge length of 1. From hex center to hex vertex is a distance of 1. From hex center to circle edge is always 1 no matter what angle. It's a circle ofc. However, if you draw a line from the center of the hex to the point halfway along the line between two vertices, that distance is less than 1. I'm not sure how much less than 1. I think the way to find that answer might be related to the equation for a circle.

If I find the relationship between the horizontal QR angle from center to hex edge and the distance to the edge at that angle, then I can use that to determine the vertical QT angle at which a ray would change from passing through the side to passing through the top when the line is at that QR angle.

**Comment by leftiness** [comment]
No. Calculating top vs side exit is not related to the equation of a circle. Think of it in terms of a triangle.

A hexagon is made up of six equilateral triangles. Point A is a vertex. Point B is the center of the hex. Point C is the next vertex over. Angle ABC is 60 degrees. Line AB is length 1. Line BC is length 1. Line CA is length 1. Equilateral ofc.

Anyway, I have to determine the relationship between the angle ABD and the length of the line BD, where D is an arbitrary point which falls along the line AC.

I know the length AB  is 1. I'll also know the angle of ABD. That's the angle on the horizontal QR plane from before. Therefore, I know the angle and the hypotenuse, and I want to know the adjacent. I can use this to find the adjacent: `cos(angle) = adjacent / hypotenuse`.

The adjacent which I've just calculated is the base of the triangle along the QT plane. It's the adjacent in the QT trig calculation. I want to know the angle. I know the adjacent. The opposite is the height difference between the top height of the hex and my rounded point at the center of the hex. That point which is at the hex edge at the top of the hex is point D along line AC from earlier. I can use this to find the vertical QT angle: `tan(angle) = opposite / adjacent`.

Lastly, I need to get the vertical QT angle of my found point. Adjacent is foundQ - roundQ. Opposite is foundT - roundT. I want the angle. I can find it in the same way: `tan(angle) = opposite / adjacent`.

So. I have the QR angle which tells me the side I'll go through if I go through the side instead of the top. That's split into 60 degree chunks. 0-60 degrees might be northeast? I'm not sure because of the way I'm rotating these triangles... Anyway. I've also got the QT angle of my found point relative to the rounded point. If that angle is less than the QT angle which would make the line exit through the top, then I exit through the side.

I should really sketch this stuff out just to be sure... And... does the fact that I've got QR instead of XY make a difference? I think QT will work okay because those are perpendicular planes, but Q isn't perpendicular to R? Does the math still work out anyway?

**Comment by leftiness** [comment]
Hm. I think I can't use QR as the horizontal angle unless I do more math to get the proper XY distance between those QR points. I can't just take the approach (foundQ - roundQ, foundR  roundR) and call that the (base, height aka adjacent, opposite).

What I can use are the hex-to-pixel and pixel-to-hex conversions on RedBlobGames that I thought I wouldn't need. See #51.

**Comment by leftiness** [comment]
It's also important to note that (after taking that step and calling it the found point) the found point might actually be equal to the round point. I need to move the found point a little further along the line in that case to ensure that I can draw a good triangle. I can do that by adding an offset. For example: found.q = (step_size.q \* 1e-6 + found.q).

Doing work like this is a real pain until I get #49 figured out.

**Comment by leftiness** [comment]
...

I may have seriously overcomplicated this. Consider an alternative to all this math.

I start at my starting point. Take a step to the first found point. Round to a round point. Subtract the previous point from the round point to compare their values. Did Q go up or down? What about SRT? Up or down? Now I know which direction I moved in. Was there a wall in the way of me taking this step? If yes, don't add this round point and break the loop. If no, add this round point and continue the loop.

...

Yeah. I think all this trig is not needed right now. The thing is that I made it very complicated by wanting to check for a wall _before_ taking a step. Then I needed to draw triangles and flatten planes and calculate angles. All of that stuff. If I check for  a wall after taking a step... then it's easy.

The trig... might... be needed when I do field of view? Maybe? Not sure. 

Also, I should do bigger feature branches. Like will I have to clean up unused code that I merged before I realized this? Not sure. Even if it's not a problem this time, it could be a problem next time if I merge to master after each little piece is completed.
